Backend/src/ml_lib/sub_models/qaqc_main_model.py

```diff
@@ -71,32 +71,8 @@ def first_run(self, df: pd.DataFrame, dim: str | None) -> pd.DataFrame:
         id_df = pd.DataFrame(data=d)
         model = SentenceTransformer("paraphrase-MiniLM-L6-v2")
         embeddings = model.encode(questions, convert_to_tensor=True)
-        match dim:
-            case "COMBO":
-                pca2 = PCA(n_components=50, whiten=False, random_state=42)
-                reduced_embeddings_pca2 = pca2.fit_transform(embeddings)
-                tsne = TSNE(n_components=2, random_state=42, perplexity=30, n_iter=300, early_exaggeration=4, learning_rate=1000)
-                reduced_embeddings = tsne.fit_transform(reduced_embeddings_pca2)
-                point_data_df = pd.DataFrame(reduced_embeddings, columns=["x", "y"])
-
-            case "PCA":
-                pca = PCA(n_components=2, random_state=42)
-                reduced_embeddings_PCA = pca.fit_transform(embeddings)
-                point_data_df = pd.DataFrame(reduced_embeddings_PCA, columns=["x", "y"])
-
-            case "TSNE":
-                pure_tsne = TSNE(n_components=2, random_state=42, perplexity=30, n_iter=300, early_exaggeration=4, learning_rate=1000)
-                reduced_embeddings_tsne = pure_tsne.fit_transform(embeddings)
-                point_data_df = pd.DataFrame(reduced_embeddings_tsne, columns=["x", "y"])
-
-            case "UMAP":
-                umap_model = umap.UMAP(n_components=2, random_state=42, n_neighbors=15, min_dist=0.1, metric='euclidean')
-                reduced_embeddings_umap = umap_model.fit_transform(embeddings)
-                point_data_df = pd.DataFrame(reduced_embeddings_umap, columns=["x", "y"])
 
-            case _: 
-                raise Exception("Invalid dimension reduction method.")
-            
+        point_data_df = self.dim_red(embeddings=embeddings, dim=dim)
```

Comment:
I think this line would be even more fitting placed in the shared models  `run` function. That way no one can accidentally forget to add the function. 

I would probably place it on line 40 in `shared_model.py`

Reply:
It's a good suggestion, however I believe that it can pose a problem with how we are handling the output from "latter_run()" at the moment. Will keep it in mind though

Reply:
I thought that both `first_run` and `latter_run` would return pandas data frames and that both would look similar with the only difference being how the nodes are clustered or placed. Is that not the case?

Reply:
That is indeed the case, however I am not quite sure how one would perform dim_red() on embeddings that havent been made yet(as this is made in the functions called from run()). I guess you could perhaps manipulate the DataFrame extracted from the first_run() or latter_run() in order to perform the embedding and dimension reduction in run(). 

However,  it may not be optimal considering that we are using a highly specific sentence transformer model(to our dataset) to extract our embeddings. There may be some obvious solution that I am not seeing, but then you'll have to illuminate me.

Reply:
Ah, I see. Well I guess it depends on exactly how the `latter_run` is implemented, so it might be a good strategy to implement that first and then see if using `dim_red` in `run` still holds. We could cycle back to this when you guys are done with the first random forest implementation.

Reply:
Sounds like a plan